### PR TITLE
Freshness revamp

### PR DESCRIPTION
**Pull request by dpeng817**
Revamps freshness checks APIs.
- Use the same code path for the implementation, making the top level functions one liners.
- Rename `maximum_lag_minutes` -> `lower_bound_delta`.
- Rename `freshness_cron` -> `deadline_cron`.
- Rename non-partitioned and time window partitioned builders to `last_update` and `time_partition` builders, respectively.

The key idea here is that if we consider all assets to have a time partition component, then we can unify the definition of freshness under the hood. An asset is considered fresh if it's "latest relevant time partition" was computed within the time window defined by some lower bound and the current time. To break it into components:
1. The existence of the latest relevant time partition
2. The latest record representing this partition being from after some lower bound

In `build_last_update_freshness_checks`, we consider the "latest relevant time partition" to be any time partition (which trivializes to no partition in the un-time-partitioned case), which means we only care about (2).

In `build_time_partition_freshness_checks`, we consider the "latest relevant time partition" to be the start of the time partition window which has completed most recently from the last tick of the cron. There is an implicit lower bound of the end of that time window which the asset's update timestamp should always be greater than, in all cases, meaning we only care about (1).

If we notice, the existence of these builders hinges on this idea of defining what the "latest relevant time partition" is. We could imagine a generalization here which allows you to define your own function for determining this time partition, to handle more complex cases such as, the current weekly partition having an update every day.

One thing worth noting is that the way we represent lower bound delta in metadata is currently using seconds, which is probably undesirable in many cases. Wondering if maybe makes sense to have a timedelta metadata param? Pretty sure TimestampMetadataValue wouldn't render properly.

How I tested this
- Existing unit tests are sufficient. They are largely unchanged outside of the two behavior changes called out earlier in this description, and param renames.
- Added a new test for complex timezone interaction between the execution timezone, the partitions timezone, and the freshness cron timezone, all of which can technically be different.

### CODE REVIEW

**Comment by dpeng817** [comment]
* **#20722** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20722?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dpeng817 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:275]: For non-partitioned assets, will the description include the word "partition"? I think we should avoid this.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:245]: What's "effective_timestamp" and where would it come from?

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:275]: no, as if the expected partition is none, we just call it "Asset"

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:245]: Something nick discussed earlier, about threading a single "effective timestamp" through the entire execution stack instead of continuously calling pendulum.now()

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py:59]: So literally the only difference here between the two functions is that once has this invariatn check?

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:240]: `effective_dt` was the terminology that was agreed upon in the `AssetGraphView` [PR](https://github.com/dagster-io/dagster/pull/20312).

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:239]: Btw this timezone only applies to the cron schedule, not the effective time, correct? Won't this cause bugs if the freshness timezone is different than the timezone of th entire system?

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py:59]: That, and the existence of the lower_bound_delta parameter on the other function.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:212]: IMO this makes things largely more confusing relative to the earlier PR.  Before, I could reason through how things worked in the different cases fairly straightforwardly. Now:
- When I'm trying to understand how freshness is defined for a non-partitioned asset, I need to reason through what partitions have to do with it (functionally: nothing).
- When I'm trying to define how freshness is defined for a partitioned asset, I need to reason through what "last update time" has to do with it (functionally: nothing).

That all said, given that this is not exposed publicly, I think the stakes are low, so I'm totally fine with shipping this version.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:239]: I guess there are actually three timezones at play here.
- The timezone of the freshness cron.
- The timezone of the partitions definition
- The timezone of the system actually running the code.
I guess what we probably need here to only track a "freshness timestamp" in terms of the actual system time, and then translate that back and forth between the various timezones. I'll add a test for this pathology.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:240]: Got it - makes sense.

That said, generally not a fan of putting TODOs in comments, because plans / names can change and easy for comments to get stale.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:270]: Ah my bad for missing.

Separate but related: the earlier version has more informative descriptions, which I think are worth preserving here.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py
Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py:59]: What is the value of `lower_bound_delta` set to here? How does this work?

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:212]: @sryza is "you" in this case the user or "you" in this case the dagster framework engineer?

Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:212]: It's a user who peeks under the hood to understand exactly what's going on.

Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:212]: and also personally literally me trying to understand what's going on in this function

Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:212]: One we move the implementation of this to use `AssetSlice` I think it will become much more clear because it will not distinguish between the partitioned and unpartitioned use case at all. It will also support dynamic, static, and multi-dimensional partitioning use cases. And it will probably be less code.

Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:212]: > I think it will become much more clear because it will not distinguish between the partitioned and unpartitioned use case at all. 

IMO this is exactly the problem. The partitioned and unpartitioned use cases are actually different here, as evidenced by the behaviors, docstrings, and descriptions. Complecting them is what makes it confusing.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:45]: type should be `timedelta`, right?

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:240]: Agreed. Will remove before we land, just thought it was helpful from a review perspective

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:45]: That's right. Oops

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py:59]: Should just be set to minutes=0. Not at PC right now but possible that bit got lost in rebase

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:21]: What timezone is the delta math done in? `timedelta(days=1)` can be 23 hours in certain time zones at certain times.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py:59]: Just to follow up, confirmed that it's included as the default timedelta.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:21]: Changed things here to make sure we keep current time a pure unix timestamp, and perform timezone conversion when necessary.

Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:21]: Just to follow up as well, I added a test for the complex timezone interaction case.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py]: Just to follow up, I added a test for the case of complex timezone interaction.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py
Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:45]: still int

**Review by schrockn — DISMISSED** [review]
Ok I made the case for a different language here in the comments around "update window" and "expiration cron". 

The other thing is that I think a scenario that will truly pressure test the generality of our mental model is the case where you say:

* Have a weekly-partitioned asset
* It is incrementally built through a week daily
* You want a *daily* freshness policy check. (say at 9 AM daily with a required update after midnight)

So in the proposed language there should be a daily expiration at 9 am, with an update window of 9 hours, applied to the latest partition.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:223]: My above comment should have been scoped to time-partitioned assets. I agree that each partition in a static-partitioned asset should normally function the same as a non-partitioned asset.

Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:223]: Got it makes sense

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:324]: I think we should be talking about "updates" instead of "records" here. There can be an observation record that doesn't indicate the asset was updated.

That said, for time-partitioned assets, I think the wording "Partition X is overdue. Expected an update" is a little misleading, because in most cases they're expecting it to be the first arrival of that partition, rather than an update to an existing partition.  The prior code uses "Partition X arrived on time" vs. "Partition X is Y minutes late", which I think avoids these difficulties.

Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:324]: > I think we should be talking about "updates" instead of "records" here. There can be an observation record that doesn't indicate the asset was updated.

Make total sense 👍🏻

At some point (outside scope of this diff) we should clarify "observation" "materialization" "record" "update" ontology and document internally and externally

Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:324]: https://linear.app/dagster-labs/issue/FOU-128/refresh-and-document-ontology-around-materialization-observation

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:30]: `update_window_duration` seems solid to me.

`expiration_cron` feels off.  Like an expiration date on a package of food, it makes me think that, as soon as the asset hits the expiration date, it's no longer considered fresh.

I think the way most users will want to think about this parameter is as a deadline.

Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:30]: Yeah def not set on expiration.

My thinking here is that if we use the "window" language we need a variable name that makes it unambiguous that it is the cron that determines the _end_ rather than the _start_ of the window. Deadline does that too IMO.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:21]: So the answer to "What timezone is the delta math done in?" is "always UTC"?

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:30]: > My thinking here is that if we use the "window" language we need a variable name that makes it unambiguous that it is the cron that determines the end rather than the start of the window

Got it - that makes sense

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:297]:
```suggestion
def get_description_for_freshness_check_result(
```

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:317]: Per [this earlier comment](https://github.com/dagster-io/dagster/pull/20722/files#r1541292406), I think we should avoid talking about "records" here.

**Review by sryza — COMMENTED** [review]
Only remaining substantive feedback here is about the wording on the description text. Other than that, I think this is good to go.

There's also of course the concerns that @schrockn raised about `lower_bound_delta`.  However @dpeng817 has made arguments that I find convincing that `lower_bound_delta` is more accurate than `update_window_duration`, so comfortable with `lower_bound_delta` if we aren't able to find time to align with Nick before we need to merge this.  Could also consider names of the form `update_age_x`.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:297]: ah yea I think i actually changed to this in the upstream. but agreed.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:174]: this should be `deadline_cron_timezone`, right?

Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:174]: Or should it actually just be`timezone`, because it also applies to the `timedelta` param?

**Review by sryza — APPROVED** [review]
Left one comment on the `freshness_cron_timezone` param, which I think needs to be renamed based on the other changes we've made.

I know the `lower_bound_delta` discussion is still ongoing, so will leave it to you to coordinate on exactly what we want to do there.

But giving this the approve because everything else looks good to go and we are a bit down to the wire here.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:174]: it should just be timezone i guess. All the relevant math is done in terms of this timezone and the result is given in terms of this timezone, I think it's fair to just call it timezone.

Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py]: 👍

**Pull request merged**
